agent/testflinger_agent/tarfile_patch.py

```diff
@@ -0,0 +1,265 @@
+
+
+class OutsideDestinationError(FilterError):
+    def __init__(self, tarinfo, path):
+        self.tarinfo = tarinfo
+        self._path = path
+        super().__init__(
+            f"{tarinfo.name!r} would be extracted to {path!r}, "
+            + "which is outside the destination"
+        )
+
+
+class SpecialFileError(FilterError):
+    def __init__(self, tarinfo):
+        self.tarinfo = tarinfo
+        super().__init__(f"{tarinfo.name!r} is a special file")
+
+
+class AbsoluteLinkError(FilterError):
+    def __init__(self, tarinfo):
+        self.tarinfo = tarinfo
+        super().__init__(f"{tarinfo.name!r} is a link to an absolute path")
+
+
+class LinkOutsideDestinationError(FilterError):
+    def __init__(self, tarinfo, path):
+        self.tarinfo = tarinfo
+        self._path = path
+        super().__init__(
+            f"{tarinfo.name!r} would link to {path!r}, "
+            + "which is outside the destination"
+        )
+
+
+def _get_filtered_attrs(member, dest_path, for_data=True):
+    new_attrs = {}
+    name = member.name
+    dest_path = os.path.realpath(dest_path)
+    # Strip leading / (tar's directory separator) from filenames.
+    # Include os.sep (target OS directory separator) as well.
+    if name.startswith(("/", os.sep)):
+        name = new_attrs["name"] = member.path.lstrip("/" + os.sep)
+    if os.path.isabs(name):
+        # Path is absolute even after stripping.
+        # For example, 'C:/foo' on Windows.
+        raise AbsolutePathError(member)
+    # Ensure we stay in the destination
+    target_path = os.path.realpath(os.path.join(dest_path, name))
+    if os.path.commonpath([target_path, dest_path]) != dest_path:
+        raise OutsideDestinationError(member, target_path)
+    # Limit permissions (no high bits, and go-w)
+    mode = member.mode
+    if mode is not None:
+        # Strip high bits & group/other write bits
+        mode = mode & 0o755
+        if for_data:
+            # For data, handle permissions & file types
+            if member.isreg() or member.islnk():
+                if not mode & 0o100:
+                    # Clear executable bits if not executable by user
+                    mode &= ~0o111
+                # Ensure owner can read & write
+                mode |= 0o600
+            elif member.isdir() or member.issym():
+                # Ignore mode for directories & symlinks
+                mode = None
+            else:
+                # Reject special files
+                raise SpecialFileError(member)
+        if mode != member.mode:
+            new_attrs["mode"] = mode
+    if for_data:
+        # Ignore ownership for 'data'
+        if member.uid is not None:
+            new_attrs["uid"] = None
+        if member.gid is not None:
+            new_attrs["gid"] = None
+        if member.uname is not None:
+            new_attrs["uname"] = None
+        if member.gname is not None:
+            new_attrs["gname"] = None
```

Comment:
I would like also to suggest the simplification here 

```suggestion
        # Ignore ownership for 'data'
        attributes = ["uid", "gid", "uname", "gname"]
        for attr in attributes:
            if getattr(member, attr) is not None:
                new_attrs[attr] = None
```

Reply:
Almost all code in the patch is copied from the standard library implementation of `tarfile`. I concede that your suggestion is equivalent, shorter and simpler but I am not too keen to make changes there. For one thing, it will remain simpler to compare versions (e.g. to verify that we don't diverge).

Reply:
Why are we copying this code instead of using it as a dependency?

Reply:
I cover that in the description of the PR: the filtering code is not available in Python 3.8.10 (which is the version the agents are running) so we provide a patch.

Reply:
Do you think we’ll remove this script and start using the lib itself after dropping python 3.8 support?

Reply:
Exactly. See also this [added comment](https://github.com/canonical/testflinger/pull/258/commits/c4c8b914c6bc2dfb73a09c372badcc9590e9b6c7) that clarifies that.